nbgitpuller/plugins/plugin_helper.py

```diff
@@ -0,0 +1,116 @@
+import subprocess
+import os
+import logging
+import requests
+from requests_file import FileAdapter
+import shutil
+import re
+
+
+# for large files from Google Drive
+def get_confirm_token(response):
+    for key, value in response.cookies.items():
+        if key.startswith('download_warning'):
+            return value
+    return None
+
+
+# sets up the a local repo that acts like a remote
+def initialize_local_repo(local_repo_path):
+    logging.info(f"Creating local_repo_path: {local_repo_path}")
+    os.makedirs(local_repo_path, exist_ok=True)
+
+    subprocess.check_output(["git", "init", "--bare"], cwd=local_repo_path)
+
+
+# local repo cloned from the "remote" which is in user drive
+def clone_local_origin_repo(origin_repo_path, temp_download_repo):
+    logging.info(f"Creating temp_download_repo: {temp_download_repo}")
+    os.makedirs(temp_download_repo, exist_ok=True)
+
+    cmd = ["git", "clone", f"file://{origin_repo_path}", temp_download_repo]
+    subprocess.check_output(cmd, cwd=temp_download_repo)
+
+
+# this is needed to unarchive various formats(eg. zip, tgz, etc)
+def determine_file_extension(url, response):
+    file_type = response.headers.get('content-type')
+    content_disposition = response.headers.get('content-disposition')
+    ext = None
+    if content_disposition:
+        fname = re.findall("filename\\*?=([^;]+)", content_disposition)
+        fname = fname[0].strip().strip('"')
+        ext = fname.split(".")[1]
+    elif file_type and "/zip" in file_type:
+        ext = "zip"
+    else:
+        url = url.split("/")[-1]
+        if "?" in url:
+            url = url[0:url.find('?')]
+        if "." in url:
+            ext = url.split(".")[1]
+
+    if not ext:
+        m = f"Could not determine the file extension for unarchiving: {url}"
+        raise Exception(m)
+    return ext
+
+
+# the downloaded content is in the response -- unarchive and save to the disk
+def save_response_content(url, response, temp_download_repo):
+    try:
+        ext = determine_file_extension(url, response)
+        CHUNK_SIZE = 32768
+        temp_download_file = f"{temp_download_repo}/download.{ext}"
+        with open(temp_download_file, "wb") as f:
+            for chunk in response.iter_content(CHUNK_SIZE):
+                # filter out keep-alive new chunks
+                if chunk:
+                    f.write(chunk)
+
+        shutil.unpack_archive(temp_download_file, temp_download_repo)
```

Comment:
This can take a while, and will block the server process. I'd suggest we shell out to `zip` or `tar` as necessary - see https://github.com/ryanlovett/jupyter-tree-download/blob/master/jupyter_tree_download/handlers.py for an example.

Reply:
yes good deal got it